sklearn/utils/__init__.py

```diff
@@ -5,25 +5,59 @@
 import platform
 import struct
 
+import warnings
 import numpy as np
 from scipy.sparse import issparse
-import warnings
+from distutils.version import LooseVersion
 
 from .murmurhash import murmurhash3_32
+from .class_weight import compute_class_weight, compute_sample_weight
+from .import _joblib
+from ..exceptions import DataConversionWarning
+from .fixes import _Sequence as Sequence
+from .deprecation import deprecated
 from .validation import (as_float_array,
                          assert_all_finite,
                          check_random_state, column_or_1d, check_array,
                          check_consistent_length, check_X_y, indexable,
                          check_symmetric)
-from .class_weight import compute_class_weight, compute_sample_weight
-from ._joblib import cpu_count, Parallel, Memory, delayed, hash
-from ._joblib import parallel_backend, register_parallel_backend
-from ._joblib import effective_n_jobs
-from ..exceptions import DataConversionWarning
-from ..utils.fixes import _Sequence as Sequence
-from .deprecation import deprecated
 from .. import get_config
 
+# deprecate the joblib API in sklearn in favor of using directly joblib
+msg = ("it is a vendored version, modified to be privately used in sklearn. "
```

Comment:
I don't get why we are doing this, but I wasn't attentive to all the past discussion. I thought deprecation here was only in order to unvendor joblib.

 If we don't unvendor joblib, we still need to provide a straightforward way for users to use parallel_backend, don't we, and that means using the vendored version and keeping at least parallel_backend public.

Reply:
That's a good point. If this method is exposed, then some users with installed `joblib` might try to use `joblib.parallel_backend`. In this case, if `SKLEARN_SITE_JOBLIB` is not set to 1, it won't have the expected behavior.
It is also true for `register_parallel_backend` and in a lesser measure for `effective_n_jobs` and `cpu_count`.

But then, would it be better to expose these functions in `sklearn.utils.joblib` or stick to `sklearn.utils`? The latter seems to make sense here as it is only for `sklearn` version of `joblib`, except if the `SKLEARN_SITE_JOBLIB` is set to 1.

Reply:
For 0.20.1 I still think we should make the minimal changes for public consistency and we can consider deprecation or unvendoring for 0.21

Reply:
In this case, I would say we should not deprecate `parallel_backend` and `register_parallel_backend` as they have direct interaction with `sklearn`. 

But we can probably deprecate the rest as it can just be used independently of `sklearn` (it is not linked to unvendoring).

Reply:
As it is highlighted in [this comment](https://github.com/scikit-learn/scikit-learn/issues/12250#issuecomment-438985467).

Reply:
But then we need to expose more functions and then deprecate them immediately, right?

Reply:
What other functions do you think we need to expose?

I think these two functions are the only one that have an impact on the behavior of scikit-learn.